Plugins/Resources/Resources.cpp

```diff
@@ -0,0 +1,282 @@
+
+const int NWSYNC_STATE_DOWNLOAD_MANIFEST_REQUEST = 10;  // We sent the manifest request.
+const int NWSYNC_STATE_DOWNLOAD_MANIFEST_PROGRESS = 11;  // Downloading the manifest.
+const int NWSYNC_STATE_DOWNLOAD_MANIFEST_PARSE = 12;  // Parsing the manifest in-memory.
+
+const int NWSYNC_STATE_WAIT_PERMISSION_TO_SYNC = 20;  // We're waiting for you/the user to give the go-ahead. Call SetSyncPermission(true) to advance.
+
+const int NWSYNC_STATE_GOT_PERMISSION_TO_SYNC = 21;  // Internal state.
+
+const int NWSYNC_STATE_SYNC_IN_PROGRESS = 100; // We're currently syncing.
+const int NWSYNC_STATE_SYNC_IN_PROGRESS_FLUSH = 101; // We're currently flushing to disk.
+
+const int NWSYNC_STATE_HOUSEKEEPING_IN_PROGRESS = 200; // Do some post-transfer cleanup.
+
+const int NWSYNC_STATE_DONE = 800; // Done! Go ahead.
+const int NWSYNC_STATE_FAILED = 801; // We aborted. Error will be in currentStateMessage.
+
+static Hooks::Hook s_loadModuleStartHook;
+static Hooks::Hook s_readFieldCExoStringHook;
+static Hooks::Hook s_GetListHook;
+
+NWNX_PLUGIN_ENTRY Plugin* PluginLoad(Services::ProxyServiceList* services)
+{
+    g_plugin = new Resources::Resources(services);
+    return g_plugin;
+}
+
+namespace Resources {
+
+Resources::Resources(Services::ProxyServiceList* services)
+: Plugin(services)
+{
+    s_loadModuleStartHook = Hooks::HookFunction(&CNWSModule::LoadModuleStart, &OnLoadModuleStart, Hooks::Order::Late);
+}
+
+uint32_t Resources::OnLoadModuleStart(CNWSModule *pModule, CExoString param_1, int param_2, int param_3, const NWSync::Advertisement &param_4)
+{
+    bool reloadRequired = false;
+    uint32_t retVal = 0;
+
+    if (Config::Get<bool>("USE_CUSTOM_HAK_LIST"))
+    {
+        retVal = g_plugin->LoadCustomHakResources();
+        if (retVal != 0)
+        {
+            return retVal;
+        }
+
+        s_GetListHook = Hooks::HookFunction(&CResGFF::GetList, &OnGetList, Hooks::Order::Late);
+        reloadRequired = true;
+    }
+
+    if (Config::Get<bool>("USE_NWSYNC"))
+    {
+        retVal = g_plugin->LoadNwSyncResources();
+        if (retVal != 0)
+        {
+            return retVal;
+        }
+
+        s_readFieldCExoStringHook = Hooks::HookFunction(&CResGFF::ReadFieldCExoString, &OnReadFieldCExoString, Hooks::Order::Late);
+        reloadRequired = true;
+    }
+
+    if (reloadRequired)
+    {
+        Globals::Rules()->ReloadAll();
+    }
+
+    return s_loadModuleStartHook->CallOriginal<uint32_t>(pModule, param_1, param_2, param_3, param_4);
+}
+
+CExoString Resources::OnReadFieldCExoString(CResGFF *pGff, CResStruct *pStructure, char *szFieldID, BOOL &bSuccess, const CExoString &sDefault)
+{
+    if (strncmp(pGff->m_pFileType, "IFO", 3) == 0 && strcmp(szFieldID, "Mod_CustomTlk") == 0)
+    {
+        if (auto tlk = g_plugin->FindNwSyncTlk())
+        {
+            LOG_INFO("Using TLK file '%s' from NWSync", tlk->CStr());
+            bSuccess = true;
+            return {*tlk};
```

Comment:
This should only be read during load, right? Would help to clear the hook afterwards, just to avoid the needless overhead in every GFF read ever afterwards.